tests/trestle/core/remote/cache_test.py

```diff
@@ -0,0 +1,371 @@
+    fetcher = cache.FetcherFactory.get_fetcher(pathlib.Path(tmp_trestle_dir), catalog_file, False, False)
+    fetcher._refresh = True
+    fetcher._cache_only = False
+    fetcher._update_cache()
+    assert fetcher._inst_cache_path.exists()
+
+
+def test_https_fetcher(tmp_trestle_dir):
+    """Test the https fetcher."""
+    os.environ['fetcher_user'] = 'fetcher'
+    os.environ['fetcher_pass'] = 'secret'
+    # This is a real, live URL:
+    uri = 'https://{{fetcher_user}}:{{fetcher_pass}}@placekitten.com/200/300'
+    fetcher = cache.FetcherFactory.get_fetcher(pathlib.Path(tmp_trestle_dir), uri, False, False)
+    fetcher._refresh = True
+    fetcher._cache_only = False
+    with patch('requests.models.Response.json') as json_mock:
+        json_mock.return_value = {'isBinary': False, 'text': str({'key': 'val'})}
+        try:
+            fetcher._update_cache()
+        except Exception:
+            AssertionError()
+
+
+def test_https_fetcher_fails(tmp_trestle_dir):
+    """Test the https fetcher failures."""
+    os.environ['fetcher_user'] = 'fetcher'
+    os.environ['fetcher_pass'] = 'secret'
+    uri = 'https://{{fetcher_user}}:{{fetcher_pass}}@placekitten.com/200/300'
+    fetcher = cache.FetcherFactory.get_fetcher(pathlib.Path(tmp_trestle_dir), uri, False, False)
+    fetcher._refresh = True
+    fetcher._cache_only = False
+    with patch('requests.auth.HTTPBasicAuth') as auth_mock:
+        auth_mock.return_value = None
+        with patch('requests.models.Response.json') as json_mock:
+            json_mock.return_value = None
+            with pytest.raises(err.TrestleError):
+                fetcher._update_cache()
+        with patch('requests.models.Response.json') as json_mock:
+            json_mock.return_value = {'isBinary': True}
+            with pytest.raises(err.TrestleError):
+                fetcher._update_cache()
+
+
+def test_sftp_fetcher(tmp_trestle_dir):
+    """Test the sftp fetcher."""
+    uri = 'sftp://some.host//path/to/test.json'
+    fetcher = cache.FetcherFactory.get_fetcher(pathlib.Path(tmp_trestle_dir), uri, False, False)
+    fetcher._refresh = True
+    fetcher._cache_only = False
+    with patch('paramiko.SSHClient.load_system_host_keys') as ssh_load_keys_mock:
+        with patch('paramiko.SSHClient.connect') as ssh_connect_mock:
+            with patch('paramiko.SSHClient.open_sftp') as sftp_open_mock:
+                with patch('paramiko.SFTPClient.get'):
+                    try:
+                        fetcher._update_cache()
+                    except Exception:
+                        AssertionError()
+                    ssh_load_keys_mock.assert_called_once()
+                    ssh_connect_mock.assert_called_once()
+                    sftp_open_mock.assert_called_once()
+
+
+def test_sftp_fetcher_cache_only(tmp_trestle_dir):
+    """Test that sftp fetcher does not call update (_sync_cache) when _cache_only is true."""
+    uri = 'sftp://some.host//path/to/test.json'
```

Comment:
There is high code repetition across these tests . . maybe create a helper function.

Reply:
I'll optimize in the end. Plenty more tests to add and update coming up.